File: api/kernels/javascript/codepod.js
// var XXX6 = 8;

// CAUTION this must be var so that it can be evaluted multiple times (during
// debugging)
var { CODEPOD_EVAL, CODEPOD_GETMOD } = (() => {
  var CODEPOD_MOD = {};

  function CODEPOD_GETMOD() {
    return CODEPOD_MOD;
  }

  function CODEPOD_EVAL(code, ns, names) {
    // ensure namespace
    if (!CODEPOD_MOD[ns]) {
      CODEPOD_MOD[ns] = {};
    }
    // console.log("=== CODEPOD: evaluate in", ns);
    // I can start to instantiate names here
    for (let k of Object.keys(CODEPOD_MOD[ns])) {
      //   console.log("=== CODEPOD: bring in", k);
      eval(`var ${k} = CODEPOD_MOD["${ns}"].${k}`);
    }
    // this is a DEF pod
    let res = eval(code);
    for (let name of names) {
      eval(`CODEPOD_MOD["${ns}"].${name} = ${name}`);
    }
    return res;
  }

  return { CODEPOD_EVAL, CODEPOD_GETMOD };
})();

// function CODEPOD3(code, ns, names) {
//   // ensure namespace
//   if (!CODEPOD_MOD[ns]) {
//     CODEPOD_MOD[ns] = {};
//   }
//   if (names && names.length > 0) {
//     // this is a DEF pod
//     console.log("CODEPOD: DEF pod");
//     let code1 = `
//         (() => {
//           // already defined names
//           for (let k of Object.keys(${ns})) {
//             eval("var \${k} = ${ns}.\${k}")
//           }
//           ${code}
//           ${ns} = {...${ns}, ${names.join(",")}}
//         })()
//         `;
//     console.log("CODE: ", code1);
//     eval(code1);
//   } else {
//     console.log("CODEPOD: EVAL pod");
//     let code1 = `
//         (() => {
//           // already defined names
//           for (let k of Object.keys(${ns})) {
//             eval("var \${k} = ${ns}.\${k}")
//           }
//           return eval("${code}")
//         })()
//         `;
//     console.log("CODE: ", code1);
//     return eval(code1);
//   }
// }

File: api/kernels/julia/codepod.jl
function isModuleDefined(names)
    mod = :(Main)
    for name in names
        name = Symbol(name)
        if !isdefined(eval(mod), name) 
            return false
        end
        mod = :($mod.$name)
    end
    return true
end

function ensureModuleDefined(namespace)
    # ensure the module is defined, and return the module
    # 1. split with /
    names = split(namespace, "/", keepempty=false)
    mod = :(Main)
    for name in names
        name = Symbol(name)
        if !isdefined(eval(mod), name) 
            include_string(eval(mod), "module $name end")
        end
        mod = :($mod.$name)
    end
    return mod, eval(mod)
end

function CODEPOD_EVAL(code, ns)
    _, mod = ensureModuleDefined(ns)
    include_string(mod, code)
end


function CODEPOD_ADD_IMPORT(from, to, name)
    @debug "ensure " from
    from_name, _ = ensureModuleDefined(from)
    @debug "ensure " to
    _, to_mod = ensureModuleDefined(to)
    newcode = "using $from_name: $name as CP$name\n$name=CP$name\n$name"
    @debug "HEBI: newcode:" newcode
    include_string(to_mod, newcode)
end

function CODEPOD_DELETE_IMPORT(ns, name)
    _, mod = ensureModuleDefined(ns)
    include_string(mod, "$name=nothing")
end